### PR TITLE
Smove 14599

### PR DESCRIPTION
**Pull request by rigazilla**
SMOVE. https://issues.redhat.com/browse/ISPN-14599
based on SINTER #11035

### CODE REVIEW

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
+         ChannelHandlerContext ctx,
+         List<byte[]> arguments) {
+      var sourceK = arguments.get(0);
+      var destinationK = arguments.get(1);
+      EmbeddedSetCache<byte[], byte[]> esc = handler.getEmbeddedSetCache();
+      CompletableFuture<SetBucket<byte[]>> sourceStg, destinationStg;
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java:37]: Why using CompletableFuture?

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      sourceStg = esc.get(sourceK).toCompletableFuture();
+      destinationStg = esc.get(destinationK).toCompletableFuture();
+      final var element = arguments.get(2);
+      AggregateCompletionStage<Void> acs = CompletionStages.aggregateCompletionStage();
+      acs.dependsOn(sourceStg).dependsOn(destinationStg);
+      final var acsFroze = acs.freeze();
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java:43]: I am really confused what is going on with this, it seems completely unneeded.

You can use `CompletionStage.thenCombine` to get a callback that is completed when both are done with their respective values.

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java [hunk 3]
```diff
@@ -44,8 +44,18 @@ public EmbeddedSetCache(Cache<K, SetBucket<V>> cache) {
     * @param key, the name of the set
     * @return the set with values if such exist, or null if the key is not present
     */
-   public CompletionStage<Set<V>> get(K key) {
-      return readWriteMap.eval(key, new SGetFunction<K, V>()).thenApply(v -> v.values());
+   public CompletionStage<SetBucket<V>> get(K key) {
```
Comment (wburns) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java:47]: We should really be doing this with new functions 
1. atomic remove which returns if the value was removed 
2. atomic add that adds the element to a set

Technically this won't make the SMOVE operation itself atomic, but at least the updates to the sets themselves can be atomic.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java (hunk 2 above)
Comment (rigazilla) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java:43]: yeah, pretty a mess. let me see

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java (hunk 3 above)
Comment (rigazilla) [reply multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java:47]: do you mean remove and add in two steps?
my intention was to do a single putAll to ensure both sets are updated in the same operation... but may this doesn't guarantee atomicity

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java (hunk 2 above)
Comment (rigazilla) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java]: updated with a new version

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java (hunk 3 above)
Comment (jabolina) [reply multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java:47]: +1 to using functions, too. Which also helps a bit in transferring the entry back and forth between nodes.

I think we can't provide an atomic operation as we don't load and lock both entries. There is always a small window when the entry will be neither on source nor destination.

Comment (rigazilla) [reply multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java:47]: yeah ok. I'll follow the lmove approach

Comment (rigazilla) [reply multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java:49]: changed the implementation to functions

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      boolean sameList = Arrays.equals(source, destination);
+      EmbeddedSetCache<byte[], byte[]> esc = handler.getEmbeddedSetCache();
+      CompletionStage<Long> resultStage = null;
+      if (!sameList) {
+         // warn when different sets
+         Log.SERVER.smoveConsistencyMessage();
+         resultStage = esc.remove(source, element).thenCompose((removed) -> removed==0 ? CompletableFuture.completedFuture(removed) : esc.add(destination, element));
+         return handler.stageToReturn(resultStage, ctx, Consumers.LONG_BICONSUMER);
+      }
+         return handler.stageToReturn(resultStage, ctx, Consumers.LONG_BICONSUMER);
```
Comment (jabolina) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java:51]: We need to do something like an exist in this case. Checking if the value exists in the source set.

File: server/resp/src/test/java/org/infinispan/server/resp/RespSetCommandsTest.java [hunk 5]
```diff
@@ -132,4 +132,28 @@ public void testSinterstore() {
...
+        assertThat(redis.sadd(dst,"2")).isEqualTo(0L);
+        assertThat(redis.sadd(src,"2")).isEqualTo(1L);
+
+        assertThat(redis.smove(src,dst,"3")).isTrue();
+        assertThat(redis.smove(src,dst,"3")).isFalse();
+
+        String nesrc = "smove-nonexist-src";
+        assertThat(redis.smove(nesrc,dst,"2")).isFalse();
+
+        String nedst = "smove-nonexist-src";
```
Comment (jabolina) [inline server/resp/src/test/java/org/infinispan/server/resp/RespSetCommandsTest.java:154]: We can extend this to cover when source == destination.

**Review by jabolina — COMMENTED** [review]
Couple of comments.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java (hunk 4 above)
Comment (rigazilla) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java]: yeah thanks, forgot about it. added

File: server/resp/src/test/java/org/infinispan/server/resp/RespSetCommandsTest.java (hunk 5 above)
Comment (rigazilla) [reply server/resp/src/test/java/org/infinispan/server/resp/RespSetCommandsTest.java]: done

**Comment by rigazilla** [comment]
rebased

File: server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      CompletionStage<Long> resultStage = null;
+      if (!sameList) {
+         // warn when different sets
+         Log.SERVER.smoveConsistencyMessage();
+         resultStage = esc.remove(source, element).thenCompose(
+               (removed) -> removed == 0 ? CompletableFuture.completedFuture(removed) : esc.add(destination, element));
+         return handler.stageToReturn(resultStage, ctx, Consumers.LONG_BICONSUMER);
+      }
+      return handler.stageToReturn(esc.get(source)
+            .thenApply((bucket) -> bucket.contains(element) ? 1L : 0L), ctx, Consumers.LONG_BICONSUMER);
```
Comment (jabolina) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/set/SMOVE.java:54]: In the future, we can update to have a contains function.

**Review by jabolina — APPROVED** [review]
LGTM, but CI is complaining about the check style.

**Comment by rigazilla** [comment]
> LGTM, but CI is complaining about the check style.

oh sorry, should be ok now

**Review by jabolina — APPROVED** [review]

**Comment by jabolina** [comment]
Thanks, @rigazilla!

**Comment by rigazilla** [comment]
thanks for the review!

**Pull request merged**